src/wallet/scriptpubkeyman.cpp

```diff
@@ -2652,6 +2652,18 @@ const std::unordered_set<CScript, SaltedSipHasher> DescriptorScriptPubKeyMan::Ge
     return script_pub_keys;
 }
 
+const std::vector<CScript> DescriptorScriptPubKeyMan::GetScriptPubKeysWithMinIndex(int32_t minimum_index) const
+{
+    LOCK(cs_desc_man);
+    std::vector<CScript> script_pub_keys;
+    script_pub_keys.reserve(m_map_script_pub_keys.size());
+
+    for (auto const& [script_pub_key, index] : m_map_script_pub_keys) {
+        if (index >= minimum_index) script_pub_keys.emplace_back(script_pub_key);
+    }
+    return script_pub_keys;
+}
```

Comment:
In an earlier version the method `GetScriptPubKeys` was indeed simply extended with a default parameter (`minimum_index = 0`). After the latest rebase however, this method is now part of the base class `ScriptPubKeyMan` interface:
https://github.com/bitcoin/bitcoin/blob/5291933fedceb9df16eb9e4627b1d7386b53ba07/src/wallet/scriptpubkeyman.h#L246
Passing a minimum index only makes sense for `DescriptorScriptPubKeyMan`s, that's why I refrained from imposing a new parameter; overloading the method _only_ for this class also seemed wrong to me, that's why I chose to use a more explicit naming. Happy to overload though if other's wish that too (maybe it's not as problematic as I think :)).

Reply:
I don't really see the issue with overloading a child class function, especially with an optional parameter? Definitely beats the code duplication in my view, but yeah would be nice to hear others' thoughts.

Reply:
There's no need to `override` it, I'm just talking about overloading it for `DescriptorScriptPubKeyMan`. For example, I think the below could work (not sure if the vector return type was required, just kept it to an unordered_set here for minimal diff):

<details>

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index f7d7f99b8..4c42ef8e0 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -2642,24 +2642,17 @@ const WalletDescriptor DescriptorScriptPubKeyMan::GetWalletDescriptor() const
 
 const std::unordered_set<CScript, SaltedSipHasher> DescriptorScriptPubKeyMan::GetScriptPubKeys() const
 {
-    LOCK(cs_desc_man);
-    std::unordered_set<CScript, SaltedSipHasher> script_pub_keys;
-    script_pub_keys.reserve(m_map_script_pub_keys.size());
-
-    for (auto const& script_pub_key: m_map_script_pub_keys) {
-        script_pub_keys.insert(script_pub_key.first);
-    }
-    return script_pub_keys;
+    return GetScriptPubKeys(0);
 }
 
-const std::vector<CScript> DescriptorScriptPubKeyMan::GetScriptPubKeysWithMinIndex(int32_t minimum_index) const
+const std::unordered_set<CScript, SaltedSipHasher> DescriptorScriptPubKeyMan::GetScriptPubKeys(int32_t minimum_index) const
 {
     LOCK(cs_desc_man);
-    std::vector<CScript> script_pub_keys;
+    std::unordered_set<CScript, SaltedSipHasher> script_pub_keys;
     script_pub_keys.reserve(std::max<size_t>(m_map_script_pub_keys.size() - minimum_index, 0));
 
     for (auto const& [script_pub_key, index] : m_map_script_pub_keys) {
-        if (index >= minimum_index) script_pub_keys.emplace_back(script_pub_key);
+        if (index >= minimum_index) script_pub_keys.insert(script_pub_key);
     }
     return script_pub_keys;
 }
diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
index ac44de0e4..4ff7fdb15 100644
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -643,7 +643,7 @@ public:
 
     const WalletDescriptor GetWalletDescriptor() const EXCLUSIVE_LOCKS_REQUIRED(cs_desc_man);
     const std::unordered_set<CScript, SaltedSipHasher> GetScriptPubKeys() const override;
-    const std::vector<CScript> GetScriptPubKeysWithMinIndex(int32_t minimum_index) const;
+    const std::unordered_set<CScript, SaltedSipHasher> GetScriptPubKeys(int32_t minimum_index) const;
 
     bool GetDescriptorString(std::string& out, const bool priv) const;
 
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index d79d9e76f..d327ab021 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -289,7 +289,7 @@ public:
         for (auto [desc_spkm, last_range_end] : m_last_range_ends) {
             int32_t current_range_end = GetDescriptorRangeEnd(*desc_spkm);
             if (current_range_end > last_range_end) {
-                for (const auto& script_pub_key : desc_spkm->GetScriptPubKeysWithMinIndex(last_range_end)) {
+                for (const auto& script_pub_key : desc_spkm->GetScriptPubKeys(last_range_end)) {
                     m_filter_set.emplace(script_pub_key.begin(), script_pub_key.end());
                 }
                 m_last_range_ends.at(desc_spkm) = current_range_end;
```

</details>

Reply:
Ah, now I fully understand what you mean, looks good at a first glance and should work. Somehow I missed the simple idea of just calling the overloaded method with from the one that is overrided (i.e. `GetScriptPubKeys(0)`) and was too focused on solving it with default parameter. Returning a `std::unordered_set` rather than `std::vector` should also be fine for our purposes. Will try that out in a bit 👍